File: data_structures_and_algorithms/04_stacks_and_queues/03_priority_queue_dynamic_typing/linked_list.py
from dataclasses import dataclass
from typing import Optional, List, Any


class NotLinkedListChild(Exception):
    pass


@dataclass
class Node:
    value: Any
    parent: 'LinkedList'
    prev: Optional['Node'] = None
    next: Optional['Node'] = None


@dataclass
class LinkedList:
    head: Optional[Node] = None
    tail: Optional[Node] = None

    def __str__(self) -> str:
        values: List[str] = []
        p = self.head
        while (p is not None):
            values.append(str(p.value))
            p = p.next

        return " -> ".join(values)

    def insert_at_end(self, new_value: Any) -> Node:
        new_node = Node(value=new_value, parent=self)

        if (self.tail is None):
            self.head = new_node
            new_node.prev = None
            self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node

        new_node.next = None

        return new_node

    def remove(self, node: Node) -> None:
        if node.parent != self:
            raise NotLinkedListChild

        prev: Optional[Node] = node.prev
        next: Optional[Node] = node.next

        if prev is None:
            self.head = next
        else:
            prev.next = next

        if next is None:
            self.tail = prev
        else:
            next.prev = prev

File: data_structures_and_algorithms/04_stacks_and_queues/02_queue_static_typing/linked_list.c
#include <stdlib.h>

typedef struct Node {
  int value;
  struct Node * prev;
  struct Node * next;
} Node;

typedef struct LinkedList {
  struct Node * head;
  struct Node * tail;
} LinkedList;

Node * insertAtEnd(LinkedList * list, int newValue) {
  if (list == NULL) return NULL;
  Node * newNode = malloc(sizeof(Node));
  newNode->value = newValue;

  if (list->tail == NULL) {
    list->head = newNode;
    newNode->prev = NULL;
    list->tail = newNode;
  }

  else {
    newNode->prev = list->tail;
    list->tail->next = newNode;
    list->tail = newNode;
  }

  newNode->next = NULL;

  return newNode;
}

Node * getFirst(LinkedList * list) {
  if (list == NULL) return NULL;
  return list->head;
}

void removeFirst(LinkedList * list) {
  if (list == NULL || list->head == NULL) return;
  Node * p = list->head;
  list->head = list->head->next;
  if (list->head != NULL) list->head->prev = NULL;
  free(p);
}
